community/index/src/main/java/org/neo4j/index/gbptree/Layout.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.index.gbptree;
+
+import java.util.Comparator;
+
+import org.neo4j.io.pagecache.PageCursor;
+
+/**
+ * Main point of interaction for customizing a {@link GBPTree}, how its keys and values are represented
+ * as bytes and what keys and values contains.
+ * <p>
+ * Additionally custom meta data can be supplied, which will be persisted in {@link GBPTree}.
+ *
+ * @param <KEY> type of key
+ * @param <VALUE> type of value
+ */
+public interface Layout<KEY,VALUE> extends Comparator<KEY>
+{
+    /**
+     * @return new key instances.
+     */
+    KEY newKey();
+
+    /**
+     * Populates {@code into} with minimum key that can ever exist in a tree with this layout.
+     *
+     * @param into instance to write minimum key into.
+     * @return the provided {@code into} instance for convenience.
+     */
+    KEY minKey( KEY into );
```

Comment:
For minKey and maxKey why do we need to supply into param? can we simply return new value with min and max key?